src/core/socket.c

```diff
@@ -519,8 +538,9 @@ int socket_acquire_peer(Socket *s, int fd, SocketPeer **ret) {
         if (!remote)
                 return log_oom();
 
-        remote->peer = sa.peer;
-        remote->peer_salen = sa.peer_salen;
+        remote->peer = key.peer;
+        remote->peer_salen = key.peer_salen;
+        remote->peer_cred = key.peer_cred;
```

Comment:
Hmm, socket_peer_new is actually only used to dup the stack-allocated SocketPeer. Hence maybe just replace that with a new function `socket_peer_dup`?